conda/common/iterators.py

```diff
@@ -18,3 +21,24 @@ def groupby_to_dict(keyfunc, sequence):
     for key, group in itertools.groupby(sequence, keyfunc):
         result[key].extend(group)
     return dict(result)
+
+
+T = TypeVar("T")
+
+
+def next_instance(objects: Iterable[Any], object_type: type[T]) -> T | None:
```

Comment:
Thanks for asking.

In type hinting, the `T` would simply be a generic for matching against an instantiated object. What we want instead is a `type` generic (that's what `type[T]` is). A `type` is precisely what `isinstance` is expecting as a second argument. 

So, in plain English the function signature translates roughly to: `object_type` should be the `type` we want to look for and then we return that instantiated type or `None`. The type that you pass in has to match the type that your return.

The Generic magic happens because it applies to multiple types (e.g. `str` -> `str`, `list` -> `list).

Hope that helps!

Reply:
Please note that `type.__getitem__` is a [Python 3.9+ functionality](https://docs.python.org/3/library/stdtypes.html#generic-alias-type).
This should still not cause trouble at runtime for 3.7/3.8 due to the [postponed evaluation](https://peps.python.org/pep-0563/) brought in via `from __future__ import annotations`.

However, beware that using features/properties of later versions can give a false sense of confirmation from a type checker. Usually it's good to also run a type checker for the lowest supported version.
(^-- That is a very general note. In this certain case it's not a big deal, probably. But if/when we have type checking as part of the CI, it will likely complain.)